### PR TITLE
Added a screen annotation tool (experimental)

### PR DESCRIPTION
**Pull request by SuhasDissa**
#### DEMO:

https://github.com/Bnyro/RecordYou/assets/64766434/6d8ff0b1-d9ae-4e11-8600-628c9665f022

### CODE REVIEW

**Comment by SuhasDissa** [comment]
Hi there!

I'm excited to announce that I've finally made the annotation tool production ready. I've tested it thoroughly, but there may still be some bugs that I haven't found.

Would you mind taking a look at it and letting me know what you think? Any suggestions are welcome.

Thanks!

**Comment by Bnyro** [comment]
This looks really cool, thanks for your work so far!
I'll take a look as soon as possible!

File: app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    private var mLifecycleRegistry: LifecycleRegistry = LifecycleRegistry(this)
+    private var mSavedStateRegistryController: SavedStateRegistryController =
+        SavedStateRegistryController.create(this)
+
+    /**
+     * @return True if the Lifecycle has been initialized.
+     */
+    val isInitialized: Boolean
+        get() = true
+
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt:24]: nit: two blank lines, please remove one

File: app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    fun setCurrentState(state: Lifecycle.State) {
+        mLifecycleRegistry.currentState = state
+    }
+
+    fun handleLifecycleEvent(event: Lifecycle.Event) {
+        mLifecycleRegistry.handleLifecycleEvent(event)
+    }
+
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt:33]: nit: two blank lines, please remove one

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.ViewModel
+
+class CanvasViewModel : ViewModel() {
+    var currentPath by mutableStateOf(PathProperties())
+    val paths = mutableStateListOf<PathProperties>()
+
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasViewModel.kt:12]: nit: please remove that line

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        try {
+            windowManager.removeView(canvasView)
+            canvasView.invalidate()
+            windowManager.removeView(toolbarView)
+            toolbarView.invalidate()
+        } catch (e: Exception) {
+            Log.e("Remove Overlay", e.toString())
+        }
+    }
+
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:105]: nit: please remove that line

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                        showCanvas()
+                    }
+                })
+            }
+        }
+    }
+
+    init {
+        val lifecycleOwner = CustomLifecycleOwner()
+        val viewModelStoreOwner = CustomViewModelStoreOwner()
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:54]:
```suggestion
       val viewModelStoreOwner = object: ViewModelStoreOwner {
            override val viewModelStore = ViewModelStore()
        }
```

File: app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+internal class CustomLifecycleOwner : SavedStateRegistryOwner {
+    private var mLifecycleRegistry: LifecycleRegistry = LifecycleRegistry(this)
+    private var mSavedStateRegistryController: SavedStateRegistryController =
+        SavedStateRegistryController.create(this)
+
+    /**
+     * @return True if the Lifecycle has been initialized.
+     */
+    val isInitialized: Boolean
+        get() = true
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt:23]: That's always true, so why do we have that? It's not used anywhere

File: app/src/main/java/com/bnyro/recorder/util/CustomViewModelStoreOwner.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.bnyro.recorder.util
+
+import androidx.lifecycle.ViewModelStore
+import androidx.lifecycle.ViewModelStoreOwner
+
+class CustomViewModelStoreOwner:ViewModelStoreOwner {
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/util/CustomViewModelStoreOwner.kt:6]: can be removed by following the comment and its only usage from above

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        toolbarView.setViewTreeViewModelStoreOwner(viewModelStoreOwner)
+        toolbarView.setViewTreeSavedStateRegistryOwner(lifecycleOwner)
+
+        hideCanvas()
+    }
+
+    fun showAll() {
+        try {
+            if (canvasView.windowToken == null) {
+                if (canvasView.parent == null) {
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:71]:
```suggestion
            if (canvasView.windowToken == null && canvasView.parent == null) {
```

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                    windowManager.addView(canvasView, params)
+                }
+            }
+            if (toolbarView.windowToken == null) {
+                if (toolbarView.parent == null) {
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:76]:
```suggestion
            if (toolbarView.windowToken == null && toolbarView.parent == null) {
```

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                    windowManager.addView(toolbarView, toolbarParams)
+                }
+            }
+        } catch (e: Exception) {
+            Log.e("Show Overlay", e.toString())
+        }
+    }
+
+    fun showCanvas() {
+        canvasView.visibility = View.VISIBLE
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:88]:
```suggestion
        canvasView.isVisible = true
```

File: app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    fun hideCanvas() {
+        canvasView.visibility = View.INVISIBLE
```
Comment (Bnyro) [inline app/src/main/java/com/bnyro/recorder/canvas_overlay/CanvasOverlay.kt:92]:
```suggestion
        canvasView.isInvisible = true
```

**Comment by Bnyro** [comment]
This works really well! Most of my comments are only some minor formatting changes.

Please also make all the strings translatable by adding them to `res/values/strings.xml` and using the `stringResource(stringId)` function instead of hardcoding them in English.

Looks great otherwise!

File: app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt (hunk 6 above)
Comment (SuhasDissa) [reply app/src/main/java/com/bnyro/recorder/util/CustomLifecycleOwner.kt:23]: I wasn't sure about that. That code is from a GitHub gist by someone else (see the link on top of that file)

**Comment by SuhasDissa** [comment]
> Please also make all the strings translatable by adding them to `res/values/strings.xml` and using the `stringResource(stringId)` function instead of hardcoding them in English.

I didn't want to mess with your strings file. I just added that strings for context. I suggest you to change those strings to more formal descriptions as you like.

**Comment by Bnyro** [comment]
Please use the `Preferences` object for handling with preferences instead of a `Context#prefereces` extension function, I'd like to keep that consistent across the whole app.
That'll be useful if we migrated to Jetpack Datastore for example.

**Comment by SuhasDissa** [comment]
Just a side note here, I suggest you to put these lines in your Preferences.kt file

```kotlin

inline fun <reified T : Enum<T>> SharedPreferences.getEnum(
    key: String,
    defaultValue: T
): T =
    getString(key, null)?.let {
        try {
            enumValueOf<T>(it)
        } catch (e: IllegalArgumentException) {
            null
        }
    } ?: defaultValue

inline fun <reified T : Enum<T>> SharedPreferences.Editor.putEnum(
    key: String,
    value: T
): SharedPreferences.Editor =
    putString(key, value.name)

val Context.preferences: SharedPreferences
    get() = getSharedPreferences("preferences", Context.MODE_PRIVATE)

@Composable
fun rememberPreference(key: String, defaultValue: Boolean): MutableState<Boolean> {
    val context = LocalContext.current
    return remember {
        mutableStatePreferenceOf(context.preferences.getBoolean(key, defaultValue)) {
            context.preferences.edit { putBoolean(key, it) }
        }
    }
}

@Composable
fun rememberPreference(key: String, defaultValue: Int): MutableState<Int> {
    val context = LocalContext.current
    return remember {
        mutableStatePreferenceOf(context.preferences.getInt(key, defaultValue)) {
            context.preferences.edit { putInt(key, it) }
        }
    }
}

@Composable
fun rememberPreference(key: String, defaultValue: String): MutableState<String> {
    val context = LocalContext.current
    return remember {
        mutableStatePreferenceOf(context.preferences.getString(key, null) ?: defaultValue) {
            context.preferences.edit { putString(key, it) }
        }
    }
}

@Composable
inline fun <reified T : Enum<T>> rememberPreference(key: String, defaultValue: T): MutableState<T> {
    val context = LocalContext.current
    return remember {
        mutableStatePreferenceOf(context.preferences.getEnum(key, defaultValue)) {
            context.preferences.edit { putEnum(key, it) }
        }
    }
}

inline fun <T> mutableStatePreferenceOf(
    value: T,
    crossinline onStructuralInequality: (newValue: T) -> Unit
) =
    mutableStateOf(
        value = value,
        policy = object : SnapshotMutationPolicy<T> {
            override fun equivalent(a: T, b: T): Boolean {
                val areEquals = a == b
                if (!areEquals) onStructuralInequality(b)
                return areEquals
            }
        })
        
  ```
  
  This makes the life much easier when it comes to managing preferences

**Comment by SuhasDissa** [comment]
> Please use the `Preferences` object for handling with preferences instead of a `Context#prefereces` extension function, I'd like to keep that consistent across the whole app. That'll be useful if we migrated to Jetpack Datastore for example.

Sorry for that. I thought you'd like that

**Comment by Bnyro** [comment]
I've seen something like that on other apps before already, and I do really like the approach.
However, in my opinion it should be used across the whole app then, and not only in a single file, which would make it very inconsistent and more difficult to maintain in general.

**Comment by Bnyro** [comment]
And our current approach should have a slightly better performance than using the extension functions because we don't recreated a `SharedPreferences` object for each preference, but that's mostly irrelevant anyways.

**Comment by SuhasDissa** [comment]
> I've seen something like that on other apps before already, and I do really like the approach. However, in my opinion it should be used across the whole app then, and not only in a single file, which would make it very inconsistent and more difficult to maintain in general.

the `rememberPreference` function lets you edit the variable directly and thus save the preference. so there's no need to save the preference each time the value change

for example

```kotlin
var checked by rememberPreference(prefKey,false)
    
Checkbox(
            checked = checked,
            onCheckedChange = {
                checked = it
            }
        )

```

**Comment by Bnyro** [comment]
> the rememberPreference function lets you edit the variable directly and thus save the preference. so there's no need to save the preference each time the value change
>
Yes, that's pretty cool!
I wouldn't mind merging a PR for that if you'd like to work on it, but then all preferences should use it the same way.

**Comment by Bnyro** [comment]
And thanks for your contribution! <3

**Pull request merged**